spec/core/v2/ics-024-host-requirements/README.md

```diff
@@ -110,11 +110,11 @@ IBC/TAO implementations MUST implement the following paths for the `provableStor
 Future paths may be used in future versions of the protocol, so the entire key-space in the provable store MUST be reserved for the IBC handler.
 
 
-| Store          | Path format                                                                | Value type        | Defined in |
-| -------------- | -------------------------------------------------------------------------- | ----------------- | ---------------------- |
-| provableStore  | "commitments/channels/{identifier}/sequences/{bigEndianUint64Sequence}"    | bytes             | [ICS 4](../ics-004-packet-semantics) |
-| provableStore  | "receipts/channels/{identifier}/sequences/{bigEndianUint64Sequence}"       | bytes             | [ICS 4](../ics-004-packet-semantics) |
-| provableStore  | "acks/channels/{identifier}/sequences/{bigEndianUint64Sequence}"           | bytes             | [ICS 4](../ics-004-packet-semantics) |
+| Store          | Path format                                       | Value type | Defined in                           |
+| -------------- | ------------------------------------------------- | ---------- | ------------------------------------ |
+| provableStore  | {channelIdentifier}|0x1|{bigEndianUint64Sequence} | bytes      | [ICS 4](../ics-004-packet-semantics) |
+| provableStore  | {channelIdentifier}|0x2|{bigEndianUint64Sequence} | bytes      | [ICS 4](../ics-004-packet-semantics) |
+| provableStore  | {channelIdentifier}|0x3|{bigEndianUint64Sequence} | bytes      | [ICS 4](../ics-004-packet-semantics) |
```

Comment:
I think you need a length prefix, or it would be vulnerable to ambiguous encoding.

Also, can't we extend the IBC client API with 3 methods with explicit parameters, so that it can verify the commitment in different ways? The IBC client API don't really need a general method to verify all possible blockchain commitments. Also, not all blockchains need to store commitments in a flattened Merkle store. They may be stored in more efficient or more secure ways, e.g. nested Merkle trees or ZK storage proofs.

For backward compatibility, we can just wrap the legacy IBC clients, and use the encoding scheme verified here to verify the IBC commitments.

Reply:
Oh right, it is not ambiguous because u64 is fixed size. Though I think it is tricky to reason about ambiguity depending on whether the subsequent fields are fixed size. It would also become ambiguous in case if we generalize sequence in the future to allow variable-length nonce bytes.

Reply:
Indeed, the one concern I had was that it could be easy to mess up in the future